### ISSUE TITLE
Release build is crashing with error code -9 after depleting entire free memory

### ISSUE THREAD

**Issue by MartinHlavna** [flutter/flutter#94461]
Hello,
this is ongoing problem and it is getting worse. Flutter build consumes a lot of memory. Problem is worse with release build. I am currently unable to build application in release mode with 15GB free ram, which is just ridiculous amount of consumed memory.

## Details

**Development OS**:  Ubuntu 20.04
**Target Platform:** Android

## Logs
```
Building without sound null safety
For more information see https://dart.dev/null-safety/unsound-null-safety

Dart snapshot generator failed with exit code -9


FAILURE: Build failed with an exception.

* Where:
Script '/home/opt/flutter-master/packages/flutter_tools/gradle/flutter.gradle' line: 1005

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/home/opt/flutter-master/bin/flutter'' finished with non-zero exit value 1
```

**Comment by imihirpaldhikar** [comment]
@MartinHlavna can you please provide some more information regarding the issue? I can confirm that this is not the problem when I am building for the release varient. My Build doesn't consume alot of memory. Further more information is needed to see what is exactly happening.

**Comment by imihirpaldhikar** [comment]
No. of pub packages shouldn't be a problem. Have you tried clearing the cache if the project and rebuilding it? One for suggestions I would give is that try once running in Android Studio by opening the ```android``` folder and try building from there. Let me know if you need help. The full logs will be helpful to debug the issue.

**Comment by maheshmnj** [comment]
Hi @MartinHlavna, Thanks for filing the issue. Please share the output of `flutter doctor -v` and the complete logs of `flutter run -v` (preferably in a text file as the output will be long).


Thank you.

**Comment by MartinHlavna** [comment]
@imihirpaldhikar I have cleaned cache multiple times, I even tried to wipe flutter sdk and download fresh zip file.
 @maheshmnj  I am attaching log files. I have also created memory consumption log 

It is just plain simple 


```watch "((date ; free -m) | tee -a memory-log.txt)```

but it may help to pinpoint when most of the memory is being depleted.

Just a quick note to avoid confusion - in logs, you may see that I have flutter in folder named ```flutter-master``` This is not master channel of flutter, it is  flutter used to build "our master" (we also have something like master and stable channel of our production apps and sometimes we also need to use older flutter for stable, before we roll flutter upgrade to production.)


[memory-log.txt](https://github.com/flutter/flutter/files/7647938/memory-log.txt)
[flutter-run-out.txt](https://github.com/flutter/flutter/files/7647939/flutter-run-out.txt)
[flutter-doctor.txt](https://github.com/flutter/flutter/files/7647940/flutter-doctor.txt)

**Comment by maheshmnj** [comment]
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.

@MartinHlavna It looks like the build is failing due to incompatibility with gradle 7.0.
Can you please try running the commands below and share the output

```
cd android
./gradlew clean
./gradlew build --warning-mode all
```

**Comment by MartinHlavna** [comment]
[gradle-clean.txt](https://github.com/flutter/flutter/files/7660550/gradle-clean.txt)
[gradle-build.txt](https://github.com/flutter/flutter/files/7660551/gradle-build.txt)

Sharing output of requested commands. 
To clarify, another computer with 32GB ram is able to build APK. I can probably measure, what peak consumption is. 

Looking at process list during build, command that consumes most RAM is:

```
/home/opt/flutter-master/bin/cache/artifacts/engine/android-arm64-profile/linux-x64/gen_snapshot --deterministic --snapshot_kind=app-aot-elf --elf=/home/martin/development/projects/HIDDEN/git-new/HIDDEN/.dart_tool/flutter_build/a8958107b8954cfb4a9e42da9ad679b5/arm64-v8a/app.so --strip /home/martin/development/projects/HIDDEN/git-new/HIDDEN/.dart_tool/flutter_build/a8958107b8954cfb4a9e42da9ad679b5/app.dill
```

**Comment by maheshmnj** [comment]
Thanks for the info @MartinHlavna, 
Leaving this issue open for further investigation from the team.

**Comment by LauDijksterhuis** [comment]
Any update on this?
Today I tried to build a profiling and release version of my app. Both failed with the dart snapshot exited with code 9 error. (Debug works fine)

After some trial and error, I found the dart code responsible for the error. Which is a huge generated file needed for one of the most important functions in our code.

**Comment by MartinHlavna** [comment]
@LauDijksterhuis 

How did you found problematic dart code? Just by removing the big file? We also have some generated files (build_runner outputs). 

Number of lines in generated files:
* 33 455 (file generated by us, basically creating instances and registering them in maps)
* 111 852  (compile time reflection using ```reflectable``` package)

**Comment by LauDijksterhuis** [comment]
Hi @MartinHlavna 

Please allow me to give you some context and explain what is happening:

### Background info
In our app there is a custom widget build with the CustomPaint widget.
This CustomPaint widget takes a CustomPainter which has over 450 different paths to paint. (The CustomPainter exists of 17217 lines of code).
The CustomPainter is modified so it provides a callback on which path has been tapped/clicked by the user.

### When the problem occurs
Running the app in debug mode on Android works fine, however when running in release/profiling mode I get the following error:

```
Dart snapshot generator failed with exit code -9


FAILURE: Build failed with an exception.

* Where:
Script '/Users/laudijksterhuis/Developer/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 1070

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/laudijksterhuis/Developer/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2m 59s
Running Gradle task 'assembleRelease'...                          181,7s
Exception: Gradle task assembleRelease failed with exit code 1
```

So I tried to pinpoint where te problem is coming from.
At first I thought the error came from Gradle, so I started changing to different Gradle versions. However, this was not the solution.

Then I wondered if the modification to the CustomPainter was the problem since the default CustomPainter has a canvas that does not provide any callback. So, I created a copy of the CustomPainter but reduced the total paths to just two. 

When I ran the app with just two paths instead of 450+, it worked fine.

So this concludes the dart code is not the problem. However, I think it might be too much for the Dart snapshot to compile to Android.

### Where the app does work
#### Android
Works in debug only

#### iOS 
Works in debug (emulator)
Profile & release not tested yet

#### Web
Works in debug and release
Profile not tested yet

### Flutter doctor -v output:
```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.0.1 21A559 darwin-x64, locale
    nl-NL)
    • Flutter version 2.8.1 at /Users/laudijksterhuis/Developer/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (5 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/laudijksterhuis/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.63.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.32.0

[✓] Connected device (1 available)
    • Chrome (web)          • chrome                    • web-javascript • Google
      Chrome 97.0.4692.71
```

## Conclusion
Reducing the total paths of the CustomPainter widget, results in a succesful build.

**Comment by LauDijksterhuis** [comment]
# Update!
I just tried to run the app on my iPad,
The same problem occurs as on Android.

Running the app in debug, works fine
Running the app in release gives error: 
```
    Dart snapshot generator failed with exit code -9
    Failed to package
    /Users/laudijksterhuis/Developer/GitHub/laudijksterhuis/countrie_counter/coun
    trie_counter.
    Command PhaseScriptExecution failed with a nonzero exit code
    note: Using new build system
    note: Planning
    note: Build preparation complete
    note: Building targets in parallel
```
Running the app in release with reduced paths (2) in CustomPainter, works fine

**Comment by LauDijksterhuis** [comment]
# Solution; what worked in my case

Because the behaviour & error were the same on iOS & Android. I tried to reduce the dart file in size.
 ## Steps I took:
1.  I renamed all variables (paths) from: `path354` to `p354`
 This was not enough
 
2. I commented out 1/5th of the code. 
 When running release, everything worked fine. I added kept adding more code to the build until it failed. (at 216 paths, +/- 11000 lines of code)

3. I split the file in to 5 smaller files (100 paths each). Which worked! 🎉  Just to be sure, I split them again in to 9 files.
![image](https://user-images.githubusercontent.com/49369680/149940086-3058d057-316e-43de-a672-9d08429b1a43.png)

    Each file has one void function that has the instructions of 50 paths. (Biggest file has 2752 lines of code)
    ![image](https://user-images.githubusercontent.com/49369680/149940188-2e41df62-fb0a-4a73-bb95-636b87d7f817.png)

    In the CustomPainter class all paths are combined by calling each void.
    ![image](https://user-images.githubusercontent.com/49369680/149939999-c2cb4616-5974-426b-b412-6963867e562c.png)

Now the app is running on Android & iOS without any problems.
And a additional benefit: the build time on iOS went down by 30 seconds (122 seconds for 216 paths in 1 file, 91 seconds for 460 paths in 9 files.) 

## Solution
Reduce the file size and split it into multiple smaller files.

I think the Dart snapshot generator has limited capacity per file. Replacing the big file with smaller ones made the build run super smooth, without speeding up my MacBook fans.

**Comment by MartinHlavna** [comment]
To elaborate on @LauDijksterhuis I have also tried to cut down generated ```reflectable```output (just minimal stub to remain compilable) and it indeed helped. Sadly we cannot split it into multiple files.

This may be indeed useful information for flutter team and help to reproduce this issue. I think it is more related to number of symbols in file rather than just number of lines. I have limited knowledge about compilers but maybe it is related to how compiler creates in memory structures representing code.

**Comment by dnfield** [comment]
Can you share a file (or project) that causes the excess memory usage?

Ideally this would be a test case we could share with the Dart compiler team and get feedback from them. It sounds like it's a case of lots of code taking lots of memory to compile.

**Comment by LauDijksterhuis** [comment]
Hi @dnfield ,

For me it was a custom version of this package:
https://github.com/simplewidgets/countries_world_map 

I wrote the package and only used chrome to run it in release mode (for the live example).
Now I use a custom version of this package in my private app. Which led to the error.

I'm not sure if the package example will give the same error (since I only tried it on Chrome so far)
But the file that crashed my app is almost identical to this file: https://github.com/simplewidgets/countries_world_map/blob/master/lib/world/simple_world/src/simple_world_painter.dart 

I hope it will help the dart team :)

**Comment by MartinHlavna** [comment]
Hi @dnfield, 

Sadly I cannot share code as it is not open source. 

I was however able to confirm error when trying to release build example from @LauDijksterhuis 's  package  
https://github.com/simplewidgets/countries_world_map

If this is not enough I could try to make dummy project to reproduce issue, but that would take some time.

**Comment by dnfield** [comment]
@a-siva and/or @alexmarkov - can you take a look at how gen_snapshot is handling https://github.com/simplewidgets/countries_world_map/blob/master/lib/world/simple_world/src/simple_world_painter.dart?

**Comment by dnfield** [comment]
@alexmarkov - I assume that there's some reason the compiler/snapshotter can't just break the method up for you?

I'm curious about how other compilers handle methods of this size.

**Comment by MartinHlavna** [comment]
@alexmarkov to elaborate more, ```reflectable``` output in our case consists of single huge ```Map``` literal which is similar case yet it is not method. It is private top-level variable. 

Is there difference between processing methods and variable initializations?

**Comment by LauDijksterhuis** [comment]
Hi @alexmarkov 
So just to clarify, my solution: https://github.com/flutter/flutter/issues/94461#issuecomment-1015386054 
In my case, is the right way to approach this?

**Comment by alexmarkov** [comment]
@dnfield Breaking methods into multiple pieces is complicated: for example, we would need to maintain stack traces as if they were not split; extracted methods should be able to use and modify local variables from the outer method, so local variables should be captured etc. This is a rather complicated machinery, and I'm not sure it makes sense to maintain that just for the sake of compiling huge methods (but this machinery might be also useful for certain code size optimizations such as outlining, so we might end up implementing it eventually).

As far as I know, in order to deal with huge functions compilers usually disable some or all optimizations or fall back to a non-optimizing mode (think of `-O0`). We also disable some of the optimizations when method is huge, but register allocation is not an optimization but an essential pass. I think it might be possible to reduce memory usage of register allocation in this case (for example, we can avoid allocating live ranges for constants if there are too many constants). Another option is to fall back to unoptimized compilation entirely. In any case, even if a huge method is compiled successfully, it will have a suboptimal code.

@MartinHlavna Initializers of static / top-level variables are also compiled as separate functions, so there's no big difference to a huge method.

@LauDijksterhuis Yes, I think your approach of splitting extremely huge methods into multiple pieces is the best way to avoid high memory usage during compilation, slow compilation and poor generated code for these methods.

**Comment by dnfield** [comment]
Would it make sense for the compiler to emit a warning when a method or initializer reaches a certain size?  Users could act on that by breaking up the method.

**Comment by rakudrama** [comment]
@LauDijksterhuis Splitting the app into more functions that are smaller will make life easier for the compiler since it is no longer compiling such large methods. However, that is just moving the lines of code around, so all the lines will still be in the app, making it a large app.

In the given example, a little bit of procedural abstraction will go a long way to massively reducing the size of the app. The pattern of code is very regular, mostly doing the same thing with different is the numbers.

The [pattern](https://github.com/simplewidgets/countries_world_map/blob/e9b9972e824a5a9742072053c1948ed511f49764/lib/countries/europe/netherlands/src/netherlands_painter.dart#L513) _moveTo-lineTo-lineTo-...-lineTo-close_ can be extracted like this:
```dart
...
    Path path2 = Path();
    path2.moveTo(size.width * 0.5276500, size.height * 0.6289382);
    path2.lineTo(size.width * 0.5269000, size.height * 0.6233372);
    path2.lineTo(size.width * 0.5280500, size.height * 0.6200700);
    path2.lineTo(size.width * 0.5289000, size.height * 0.6187865);
    path2.lineTo(size.width * 0.5299500, size.height * 0.6213536);
    path2.lineTo(size.width * 0.5289500, size.height * 0.6229872);
    path2.lineTo(size.width * 0.5284500, size.height * 0.6248541);
    path2.lineTo(size.width * 0.5283500, size.height * 0.6281214);
    path2.lineTo(size.width * 0.5276500, size.height * 0.6289382);
    path2.close();
...
```
⟹
```dart
...
  Path path2 = Path();
  addLines(path2, size.width, size.height, const [
     0.5276500, 0.6289382,
     0.5269000, 0.6233372,
     0.5280500, 0.6200700,
     0.5289000, 0.6187865,
     0.5299500, 0.6213536,
     0.5289500, 0.6229872,
     0.5284500, 0.6248541,
     0.5283500, 0.6281214,
     0.5276500, 0.6289382]);
...
// shared helper
Path addLines(Path path, double width, double height, List<double> points) {
  path.moveTo(width * points[0], height * points[1]);
  for (int i = 2; i + 1 < points.length; i += 2) {
    path.lineTo(width * points[i], height * points[i + 1]);
  }
  return path..close();
}
...
```
Note that the list of numbers can be a **const** list, so that it is created by the compiler and does not need to be constructed at run-time. Without the __const__, there would be compiled code in the function to create the list and fill it, having a similar size (proportional to the number of points rather than the number of segments) to the original program.

/cc @lrhn const typed data would be helpful here.

**Comment by mraleph** [comment]
I think this sort of huge functions should almost always be represented by some kind of _intermediate (byte)code_. It is will improve compilation speed, but also produce smaller and (potentially) faster binaries if interpreter is written well. 

@MartinHlavna if you have a huge map produced by `reflectable` tripping the compiler you likely have a cascade of other issues with the size (and potentially) performance of your application. How many classes do you reflect on?

**Comment by LauDijksterhuis** [comment]
@rakudrama 

Wow thank you for pointing that out, I have been looking for ways to optimize this package.
And I have been experimenting (on a much smaller scale) to dynamically load the paths into the painter via an API without losing the ability to touch countries & getting a callback.

I will definitely use this in a future release of the package 😃 👍🏼

**Comment by MartinHlavna** [comment]
@alexmarkov Thanks for reply I thought so. 

@mraleph We have one project with shared code that contains 140 reflected objects (this is essentially our platform). This project is imported into other projects (as package from git) that can extend core objects and/or provide it's own classes. These "implementation" projects are the ones that end up generating reflection. Each project typically contains about 20-30 more reflected objects.  

So you can say our "worst case" scenario is still under 200 reflected classes.  

Yes generated APK has about 100MB and release build can take some minutes to finish (about 7-10 mins, but that includes ```build_runner``` execution). We are not noticing any runtime performance problems though.

**Comment by darshankawar** [comment]
@MartinHlavna 
Revisiting this issue to reflect correct status, just would like to know if there's a way for you to know if the originally reported crash still occurs using latest framework version ?
